src/java.desktop/macosx/native/libawt_lwawt/awt/AWTWindow.m

```diff
@@ -841,7 +841,7 @@ - (void) activateWindowMenuBar {
         isDisabled = !awtWindow.isEnabled;
     }
 
-    if (menuBar == nil) {
+    if (menuBar == nil && [ApplicationDelegate sharedDelegate] != nil) {
```

Comment:
Can you please provide details on how the delegate can be null here and there if it should be set at the startup of any awt application?

Reply:
It can be null if the NSApplication was created by code outside of AWT. Specifically, in the case of an FX or SWT toolkit, but in theory a native application using JAWT could do this.

In practice, since we haven't even seen a crash as a result of this, I doubt that this code is ever called if AWT isn't running the event loop and hasn't installed their delegate.

Reply:
To access any native code in AWT the app should initially init the LWCToolkit which sets that application delegate if it is not set already. Do we know how the FX can bypass that initialization?

Reply:
In the case when AWT is embedded AWT does NOT start an  NSApplication and does not own it.
AWT overwriting FX's delegate, or any one else's is just plain wrong.

We can turn around what you said
"To access any native code in JavafX, the app should initialise GlassToolkit which sets its delegate, if not already set".

But you can't have both of these.

Meaning AWT isn't special here. It has to play by the same rules.

The toolkit that starts the NSApplication is the one that should be setting the delegate.
If you are embedded in someone else's application, it seems most logical that you defer to their application
and if this means some things aren't possible,  so be it.

Reply:
> Can you please provide details on how the delegate can be null here and there if it should be set at the startup of any awt application?

I commented on these in https://bugs.openjdk.org/browse/JDK-8319255 which is now closed as a dup.
There I wrote (in part) 
----

I am not sure if these are all cases that can provably be in un-reachable code in the "nil" case, which
means some embedded case. But if we ever do reach them, it seems we'd crash the app.
And if we need to make the delegate "nil" in the non-subclassed NSApplication case (ie like FX)
then we are more likely to cause a crash.

[list of places elided]

All these call sites need to be examined to understand that, and if necessary a test developed to
confirm it.
And even then, it might be prudent to add checks for nil.
----

So the request was to investigate and understand if they were needed, not just add them
@honkar-jdk can comment on what she did there.

Reply:
@mrserb Based on testing so far, there hasn't been a scenario where the shared delegate has been null even in  FX embedded case. As @kevinrushforth mentioned I was thinking of the case where NSApp is created by code outside of AWT and perhaps it fails to install its delegate. The present null checks were added keeping the above scenario in mind.

I'll do some more testing and get back to you on your question of - _"if shared delegate can be null at the above locations"_

Reply:
@mrserb Shared delegate can be null if it is not installed when one of following conditions is met [ApplicationDelegate.m](https://github.com/openjdk/jdk/blob/5493f5f92d569d8e94d1f271480f11c48257e896/src/java.desktop/macosx/native/libawt_lwawt/awt/ApplicationDelegate.m#L118C4-L128C33) which is currently the case for Swing components embedded in FX window. 

With https://github.com/openjdk/jfx/pull/1280 JavaFX NSApp is no longer of type NSApp, it is of type NSAppFX and considering the scenario - Swing components embedded in FX window, we hit the case where sharedDelegate is null at places where null checks have been added.

```
BOOL shouldInstall = NO;
BOOL overrideDelegate = (getenv("AWT_OVERRIDE_NSDELEGATE") != NULL);
if (NSApp != nil) {
    if ([NSApp isMemberOfClass:[NSApplication class]] && overrideDelegate) shouldInstall = YES;
    if ([NSApp isKindOfClass:[NSApplicationAWT class]]) shouldInstall = YES;
}
checked = YES;
if (!shouldInstall) return nil;

sApplicationDelegate = [[ApplicationDelegate alloc] init];
return sApplicationDelegate;
```

Reply:
@prrace @kevinrushforth @mrserb 

The reason has to why we don't observe a NPE or crash when sharedDelegate is null is because in Objective C when a method is called on a nil object it is treated as no-op and returns nil instead of NPE, contrary to what we would usually expect. Details [here](https://stackoverflow.com/questions/2696891/calling-a-method-on-an-uninitialized-object-null-pointer).

For example executing the following line when shared delegate is null and accessing defaultMenuBar on null delegate returns null and not NPE.

`[[ApplicationDelegate sharedDelegate] defaultMenuBar];`

Additionally we are not seeing any difference in defaultMenuBar with unpatched JDK + unpatched JFX (Non-Nil shared delegate) vs patched JDK + patched JFX (Nil shared delegate) at [AWTWindow.m](https://github.com/openjdk/jdk/blob/5493f5f92d569d8e94d1f271480f11c48257e896/src/java.desktop/macosx/native/libawt_lwawt/awt/AWTWindow.m#L844) and [CMenubar.m](https://github.com/openjdk/jdk/blob/5493f5f92d569d8e94d1f271480f11c48257e896/src/java.desktop/macosx/native/libawt_lwawt/awt/CMenuBar.m#L213)  because in both cases defaultMenuBar returned is nil. This might be the case at other places too and are being checked if they worked differently without and with this JDK patch.

Reply:
So in  either case -- with or without the delegate being overridden -- the defaultMenuBar is returning `nil`. This makes sense, since in the case of a JavaFX app, JavaFX has already initialized the `NSApplication` and set the system menu bar.

I do recommend leaving the `sharedDelegate` null checks that you added as part of this PR, since that way we short-circuit any other logic (e.g., in this specific case you avoid setting `isDisabled = NO` which seems more accurate), and you aren't relying on Objective C to treat this case as a no-op.

You might even consider making the `sharedDelegate` null check (and bailing out) earlier, but maybe there is a good reason to leave the check where it is.

Reply:
>The reason has to why we don't observe NPE or crash when sharedDelegate is null is

But why it is null? The OSXAPP_SetApplicationDelegate is called when the LWCToolkit is initialized in non-headless [mode](https://github.com/openjdk/jdk/blob/9a6ca233c7e91ffa2ce9451568b3be88ccd04504/src/java.desktop/macosx/native/libawt_lwawt/awt/LWCToolkit.m#L355) and that code does not depend on "NSApplicationAWT". Does it mean that the toolkit is not initiated? Then how we can call peer-menu-related code?

Reply:
I found a reason in the one comment above.

Reply:
I don't think any further changes are needed to this PR. It's already the case that if another toolkit or the app itself initializes the NSApplication, certain things won't be available in AWT. The system menu bar for example will belong to the toolkit that initializes the NSApplication. That was true before this change and it is still true.

Reply:
@mrserb 

The question as to whether sharedDelegate can be null is answered as part of https://github.com/openjdk/jdk/pull/16569#discussion_r1399559545.

> I guess the code [ApplicationDelegate.m](https://github.com/openjdk/jdk/blob/5493f5f92d569d8e94d1f271480f11c48257e896/src/java.desktop/macosx/native/libawt_lwawt/awt/ApplicationDelegate.m#L118C4-L128C33) should be changed. Right now it has some checks to prevent "double initialization", but it should take care about the case if the delegate is set already by some external lib.
Leaving that as null and skip it in the code means that part of the API will stop working, mostly java.awt.desktop package?

As Kevin explained it earlier, it would be the responsibility of the toolkit that initialized NSApp to add system menu bar, dock menu (if any). This has been tested using variations of sample programs added to [JDK-8318854](https://bugs.openjdk.org/browse/JDK-8318854).